---
title: "Poster Data Analysis"
author: "Håkon Kaspersen"
date: '`r format(Sys.time(), "%d.%m.%Y")`'
output:
    html_document:
      code_folding: hide
      df_print: paged
      highlight: kate
      number_sections: yes
      theme: flatly
      toc: yes
      toc_depth: 3
      toc_float:
        collapsed: yes
    pdf_document:
      toc: yes
      toc_depth: '3'
---

# Description
This notebook presents the analysis of whole genome data for the poster "Quinolone resistance mechanisms found in E. coli from four animal species in Norway
", presented at the ASM NGS Conference in Washington D.C. September 2018.

Please press the "code" button for each chunk to view code.
```{r message=FALSE, warning=FALSE}
# Libraries
if (!require("pacman")) install.packages("pacman")
pacman::p_load(dplyr, ggplot2, tidyr, cowplot, viridis, forcats, radiant.data, purrr, phangorn, gridExtra, ggtree)
```

# Functions
This chunk lists the functions used in this notebook.
```{r}
# Functions

func_paste <- function(x) paste(unique(x[!is.na(x)]), collapse = ", ")

## COnfidence intervals
get_binCI <- function(x, n) as.numeric(setNames(binom.test(x,n)$conf.int*100,
                                                c("lwr", "upr")))

# Creates a correlation matrix from 1/0 data from MIC values
create_corr_matrix <- function(df) {
  df <- df %>%
    as.matrix %>%
    cor %>%
    as.data.frame %>%
    rownames_to_column(var = "var1") %>%
    gather(var2, value, -var1) %>%
    mutate(var1 = gsub("(.*?)_R$", "\\1", var1),
           var2 = gsub("(.*?)_R$", "\\1", var2))
  return(df)
}

# Creates a data frame with p-values for comparing presence of specific genes between animal species
do_chisq_test <- function(df) {
  df <- df %>%
    slice(expand.grid(1:length(unique(species)), 1:length(unique(species))) %>% rev %>% 
          filter(Var2 < Var1) %>% t) %>%
    mutate(test = rep(1:(n()/2), each = 2)) %>%
    group_by(test) %>%
    do(data_frame(gene = .$key,
                  test = first(.$test),
                  species1 = first(.$species),
                  species2 = last(.$species),
                  data = list(matrix(c(.$Absent,
                                       .$Present),
                                     ncol = 2)))) %>%
    mutate(chi_test = map(data, chisq.test, correct = FALSE)) %>%
    mutate(p.value = map_dbl(chi_test, "p.value"),
           dupl = duplicated(test)) %>%
    filter(dupl == FALSE) %>%
    select(gene, species1, species2, p.value) %>%
    filter(p.value < 0.05)
  return(df)
}

"%not_in%" <- Negate("%in%")

```

# Import data
This chunk imports data from MLST and ARIBA[1] AMR gene detection results. Isolate data is also imported.
Four isolates were excluded due to bad sequence quality.
```{r}
# Isolate data, including DNA extraction results, species,
# method of isolation, and sequencing methodology
isolate_data <- read.table("../data/isolate_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% c("2016-02-428-2-S","2016-02-486-2-S","2016-02-732-2-S","2016-17-565-1-S"))

# Total MIC profile for all isolates in this study
mic_data <- read.table("../data/mic_data.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% c("2016-02-428-2-S","2016-02-486-2-S","2016-02-732-2-S","2016-17-565-1-S"))

# ARIBA MLST results for each isolate
mlst_results <- read.table("../data/mlst_results.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% c("2016-02-428-2-S","2016-02-486-2-S","2016-02-732-2-S","2016-17-565-1-S"))

# ARIBA AMR gene results (megares database, for mutations in intrinsic genes)
megares_results <- read.table("../data/megares_results.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  select(-acrB) %>%
  filter(id %not_in% c("2016-02-428-2-S","2016-02-486-2-S","2016-02-732-2-S","2016-17-565-1-S"))

# ARIBA AMR gene results (resfinder database, for acquired genes)
resfinder_results <- read.table("../data/resfinder_results.txt",
                           sep = "\t",
                           header = T,
                           stringsAsFactors = F) %>%
  filter(id %not_in% c("2016-02-428-2-S","2016-02-486-2-S","2016-02-732-2-S","2016-17-565-1-S"))

species_data <- select(isolate_data, c(id, species))
```

# Tables
## Samples per species
```{r}
per_species <- isolate_data %>%
  group_by(species) %>%
  count()
per_species
```

## Samples per sequencing method
The isolates were sequenced by three different methods due to being part of different projects. The table below lists each sequencer used, library prep, and the resulting read length.
```{r}
per_seq <- isolate_data %>%
  group_by(sequencer, libprep, read_length) %>%
  count() %>%
  rename("Sequencer" = sequencer,
         "Library Prep" = libprep,
         "Read Length" = read_length)
per_seq
```

## Total percentage of isolates with detected acquired genes and mutant intrinsic genes

### Percent acquired genes
The total percentage of presence of acquired genes identified with ARIBA.
```{r}
resfinder_results %>%
  select(-c(id, type)) %>%
  mutate(total = rowSums(.),
         result = if_else(total != 0, 1, 0)) %>%
  group_by(result) %>%
  count() %>%
  ungroup() %>%
  mutate(result = if_else(result == 0, "Absent", "Present")) %>%
  spread(result, n) %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total * 100,2),
         lwr = round(get_binCI(Present, Total)[1], 2),
         upr = round(get_binCI(Present, Total)[2], 2),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr))
```

### Percent mutations in intrinsic genes
The total percentage of mutations in intrinsic genes identified with ARIBA.
```{r}
megares_results %>%
  select(-c(id, type)) %>%
  mutate(total = rowSums(., na.rm = TRUE),
         result = if_else(total != 0, 1, 0)) %>%
  group_by(result) %>%
  count() %>%
  ungroup() %>%
  mutate(result = if_else(result == 0, "Absent", "Present")) %>%
  spread(result, n) %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total * 100,2),
         lwr = round(get_binCI(Present, Total)[1], 2),
         upr = round(get_binCI(Present, Total)[2], 2),
         "95 % CI" = paste("[", lwr, "-", upr, "]")) %>%
  select(-c(lwr, upr))
```

## MLST Data
### Total MLST
The total amount of sequence types identified.
```{r}
total_mlst <- mlst_results %>%
  left_join(species_data, by = "id") %>%
  group_by(ST) %>%
  count()
total_mlst
```

### MLST per animal species
Sequence types for each animal species.
```{r}
mlst_per_species <- mlst_results %>%
  left_join(species_data, by = "id") %>%
  group_by(ST, species) %>%
  count()
mlst_per_species
```

### Number of sequence types per animal species
The total amount of unique sequence types per animal species.
```{r}
total_ST_per_species <- mlst_per_species %>%
  group_by(species) %>%
  count()
total_ST_per_species
```

# Figures
## Total MLST
This figure presents the amount of isolates of each sequence type in total.
```{r}
mlst_results %>%
  left_join(species_data, by = "id") %>%
  group_by(ST) %>%
  count() %>%
  ggplot(aes(reorder(ST, -n), n))+
  geom_col(color = "black")+
  labs(x = "Sequence Types")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, size = 6, hjust = 1, vjust = 0.3))
```

## Percent presence of acquired genes in total
This plot presents the total percent occurrence of acquired genes identified with ARIBA, with 95 % confidence intervals.
```{r}
resfinder_results %>%
  gather(gene, result, -c(id, type)) %>%
  group_by(gene, result) %>%
  count() %>%
  ungroup() %>%
  mutate(result = if_else(result == 1, "Present", "Absent")) %>%
  spread(result, n, fill = 0) %>%
  rowwise() %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total*100, 1),
         lwr = round(get_binCI(Present, Total)[1], 1),
         upr = round(get_binCI(Present, Total)[2], 1)) %>%
  ggplot(aes(gene, Percent))+
  geom_col(color = "black")+
  geom_errorbar(aes(ymin = lwr, ymax = upr),
                width = 0.6)+
  labs(y = "Percent (%) of isolates")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, size = 8, hjust = 1, vjust = 0.4),
        axis.title.x = element_blank())
```

## Percent presence of mutations in intrinsic genes in total
This plot presents the total percent occurrence of mutations identified with ARIBA in intrinsic genes, with 95 % confidence intervals.
```{r}
megares_results %>%
  gather(gene, result, -c(id, type)) %>%
  group_by(gene, result) %>%
  count() %>%
  ungroup() %>%
  mutate(compl = complete.cases(.),
         result = if_else(result == 1, "Present", "Absent")) %>%
  filter(compl == TRUE) %>%
  select(-compl) %>%
  spread(result, n, fill = 0) %>%
  rowwise() %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total*100, 1),
         lwr = round(get_binCI(Present, Total)[1], 1),
         upr = round(get_binCI(Present, Total)[2], 1)) %>%
  ggplot(aes(gene, Percent))+
  geom_col(color = "black")+
  geom_errorbar(aes(ymin = lwr, ymax = upr),
                width = 0.6)+
  labs(y = "Percent (%) of isolates")+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, size = 8, hjust = 1, vjust = 0.4),
        axis.title.x = element_blank())
```

## Acquired genes per animal species
### Total acquired genes per animal species
This plot presents the percent occurrence of acquired genes identified with ARIBA, with 95 % confidence intervals. Please note that not only quinolone resistance mechanisms is included in this plot. Colors: ColorBrewer[2]
```{r}
gene_names <- names(resfinder_results)

gene_names <- gene_names[-c(1,2)]

resfinder_results %>%
  left_join(isolate_data, by = "id") %>%
  gather(key, value, gene_names) %>%
  group_by(key, value, species) %>%
  count() %>%
  ungroup() %>%
  mutate(value = if_else(value == 1, "Present", "Absent")) %>%
  spread(value, n, fill = 0) %>%
  rowwise() %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total*100, 1),
         lwr = round(get_binCI(Present, Total)[1], 1),
         upr = round(get_binCI(Present, Total)[2], 1)) %>%
  ggplot(aes(key, Percent, fill = species))+
  geom_col(color = "black",
           position = position_dodge(1))+
  geom_errorbar(aes(ymin = lwr, ymax = upr),
                width = 0.6,
                position = position_dodge(1))+
  scale_fill_brewer(palette = "Paired")+
  labs(y = "Percent (%) of isolates")+
  guides(fill = FALSE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, size = 8, hjust = 1, vjust = 0.4),
        axis.title.x = element_blank())+
  facet_wrap(~species)
```

### Qnr-genes per animal species
This plot presents the percent occurrence of qnr-genes identified with ARIBA in the four animal species, with 95 % confidence intervals.
```{r}
qnr_genes <- c("qnrA1","qnrB19","qnrB6","qnrB60","qnrS1","qnrS2","qnrS4")

qnr_data <- resfinder_results %>%
  left_join(isolate_data, by = "id") %>%
  gather(key, value, qnr_genes) %>%
  group_by(key, value, species) %>%
  count() %>%
  ungroup() %>%
  mutate(value = if_else(value == 1, "Present", "Absent")) %>%
  spread(value, n, fill = 0) %>%
  rowwise() %>%
  mutate(Total = Present + Absent,
         Percent = round(Present/Total*100, 1),
         lwr = round(get_binCI(Present, Total)[1], 1),
         upr = round(get_binCI(Present, Total)[2], 1))

ggplot(qnr_data, aes(key, Percent, fill = species))+
  geom_col(color = "black",
           position = position_dodge(0.9))+
  geom_errorbar(aes(ymin = lwr, ymax = upr),
                width = 0.6,
                position = position_dodge(0.9))+
  scale_fill_brewer(palette = "Paired")+
  labs(y = "Percent (%) of isolates",
       fill = NULL)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, size = 8, hjust = 1, vjust = 0.4),
        axis.title.x = element_blank())
```

## Mutations in intrinsic genes per animal species
This plot presents the percent occurrence of mutated intrinsic genes identified with ARIBA per animal species, with 95 % confidence intervals.
```{r}
mut_names <- names(megares_results)

mut_names <- mut_names[-c(1,2)]

megares_results %>%
  left_join(isolate_data, by = "id") %>%
  gather(key, value, mut_names) %>%
  group_by(key, value, species) %>%
  count() %>%
  ungroup() %>%
  mutate(value = if_else(value == 1, "Mutant", "Wild Type"),
         compl = complete.cases(.)) %>%
  filter(compl == TRUE) %>%
  spread(value, n, fill = 0) %>%
  rowwise() %>%
  mutate(Total = Mutant + `Wild Type`,
         Percent = round(Mutant/Total*100, 1),
         lwr = round(get_binCI(Mutant, Total)[1], 1),
         upr = round(get_binCI(Mutant, Total)[2], 1)) %>%
  ggplot(aes(key, Percent, fill = species))+
  geom_col(color = "black",
           position = position_dodge(1))+
  geom_errorbar(aes(ymin = lwr, ymax = upr),
                width = 0.6,
                position = position_dodge(1))+
  scale_fill_brewer(palette = "Paired")+
  labs(y = "Percent (%) of isolates")+
  guides(fill = FALSE)+
  theme_classic()+
  theme(axis.text.x = element_text(angle = 90, size = 8, hjust = 1, vjust = 0.4),
        axis.title.x = element_blank())+
  facet_wrap(~species)
```

## Correlation plots
These plots present the $\phi$-coefficient for the association between the presence/absence of mutations in intrinsic genes, and the presence/absence of acquired genes.
### Mutations in intrinsic genes
```{r}
corr_megares <- megares_results %>%
  select(-type) %>%
  left_join(species_data, by = "id") %>%
  mutate(compl = complete.cases(.)) %>%
  filter(compl == TRUE) %>%
  split(., f = .$species) %>%
  lapply(., function(x) select(x, -c(species, id, compl)))

mut_corr_matrices <- lapply(corr_megares, function(x) create_corr_matrix(x)) %>%
  bind_rows(.id = "species")

cols <- c("#e7f0fa","#c9e2f6","#95cbee","#0099dc","#4ab04a",
          "#ffd73e","#eec73a","#e29421","#f05336","#ce472e")

ggplot(mut_corr_matrices, aes(var1, var2, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradientn(colors=cols,
                       na.value="grey95",
                       limits=c(-1, 1),
                       guide=guide_colourbar(ticks=T,nbin=10,
                                             barheight=.5,label=T, 
                                             barwidth=10))+
  labs(fill = NULL)+
  theme_classic()+
  theme(axis.text = element_text(size = 7),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.4),
        axis.title = element_blank(),
        legend.position="bottom",
        legend.justification="center",
        legend.direction="horizontal")+
  coord_fixed()+
  facet_wrap(~species)
```

### Presence/absence of acquired genes
```{r message=FALSE, warning=FALSE}
corr_resfinder <- resfinder_results %>%
  select(-type) %>%
  left_join(species_data, by = "id") %>%
  split(., f = .$species) %>%
  lapply(., function(x) select(x, -c(species, id)))

acquired_corr_matrices <- lapply(corr_resfinder, function(x) create_corr_matrix(x)) %>%
  bind_rows(.id = "species")

cols <- c("#e7f0fa","#c9e2f6","#95cbee","#0099dc","#4ab04a",
          "#ffd73e","#eec73a","#e29421","#f05336","#ce472e")

ggplot(acquired_corr_matrices, aes(var1, var2, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradientn(colors=cols,
                       na.value="grey95",
                       limits=c(-1, 1),
                       guide=guide_colourbar(ticks=T,nbin=10,
                                             barheight=.5,label=T, 
                                             barwidth=10))+
  labs(fill = NULL)+
  theme_classic()+
  theme(axis.text = element_text(size = 6),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.4),
        axis.title = element_blank(),
        legend.position="bottom",
        legend.justification="center",
        legend.direction="horizontal")+
  coord_fixed()+
  facet_wrap(~species)
```

## MIC plots
### MIC Distribution for generated groups
This plot represents the MIC distribution for each auto-generated group of acquired genes and mutations combination. The plot on the bottom represents the group on the x-axis, and the acquired genes (top part of the horizontal line) and mutations in intrinsic genes (bottom part) that is present in the given group. The middle plot represents the MIC-distribution for the given group, where the size of the dot represents the amount of isolates at that MIC-value. Finally, the top plot represents the amount of isolates in the given group. All three plots have the same x-axis.
```{r message=FALSE, warning=FALSE}
selected_genes <- c("qnrA1","qnrB19","qnrB6","qnrB60","qnrS1","qnrS2","qnrS4")
selected_mut <- c("gyrA","gyrB","parC","parE","marR")

micplot_data <- mic_data %>%
  select(id, CIP) %>%
  left_join(megares_results, by = "id") %>%
  left_join(resfinder_results, by = "id") %>%
  select(-c(type.x, type.y)) %>%
  mutate(compl = complete.cases(.)) %>%
  filter(compl == TRUE) %>%
  select_at(vars(id, CIP,selected_mut, selected_genes)) %>%
  mutate(none = if_else(rowSums(.[,3:ncol(.)]) != 0, 0, 1)) %>%
  group_by_at(vars(selected_mut, selected_genes, none)) %>%
  { mutate(ungroup(.), group = group_indices(.)) } %>%
  gather(gene, value, -c(id, CIP, group)) %>%
  mutate(gene = ifelse(value == 0 & gene != "none", NA, gene),
         value = ifelse(gene == "none" & value == 0, NA, value)) %>%
  na.omit()

factor_breaks <- c(0.015, 0.03, 0.06, 0.12, 0.25, 0.5, 1, 2, 4, 8, 16)
  
factor_levels <- c("0.015", "0.03", "0.06", "0.12", "0.25", "0.5", "1", "2", "4", 
                   "8", "16")

p0 <- micplot_data %>%
  group_by(id, group) %>%
  count() %>%
  group_by(group) %>%
  count() %>%
  ggplot(aes(factor(group), nn, fill = factor(group)))+
  geom_col(color = "black")+
  guides(fill = FALSE)+
  scale_fill_viridis(discrete = TRUE,
                     direction = -1)+
  theme_light()+
  theme(axis.title = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid = element_blank())

p1 <- micplot_data %>%
  group_by(CIP, group, gene) %>%
  count() %>%
  ggplot(aes(factor(group), CIP, fill = factor(group), size = n))+
  geom_point(pch = 21)+
  scale_y_continuous(trans = "log10",
                     labels = factor_levels,
                     breaks = factor_breaks)+
  guides(fill = FALSE,
         size = FALSE)+
  scale_fill_viridis(discrete = TRUE,
                     direction = -1)+
  theme_light()+
  theme(axis.title = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.text.x = element_blank(),
        axis.ticks.x = element_blank(),
        panel.grid = element_blank())

p2 <- ggplot(micplot_data,
       aes(factor(group),
           factor(gene),
           fill = factor(group)))+
  geom_point(shape = 23,
             pch = 21,
             size = 2)+
  geom_hline(yintercept = 5.5)+
  guides(fill = FALSE)+
  scale_fill_viridis(discrete = TRUE,
                     direction = -1)+
  labs(y = "Genes")+
  theme_light()+
  theme(axis.title = element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_text(size = 8),
        axis.ticks.x = element_blank(),
        panel.grid = element_blank())

plot_grid(p0, p1, p2, align = "v", nrow = 3, rel_heights = c(1/4, 2/4, 1/4))
```

### MIC correlation plots per animal species
This plot presents the $\phi$-coefficient for correlation between 1/0 values created from the MIC-values (1 = resistant, 0 = nonresistant).
```{r}
mic_per_species <- mic_data %>%
  select(-antres) %>%
  left_join(species_data, by = "id") %>%
  mutate(SMX_R = ifelse(SMX > 64, 1, 0),
         TMP_R = ifelse(TMP > 2, 1, 0),
         CIP_R = ifelse(CIP > 0.06, 1, 0),
         TET_R = ifelse(TET > 8, 1, 0),
         NAL_R = ifelse(NAL > 16, 1, 0),
         CTX_R = ifelse(CTX > 0.25, 1, 0),
         CHL_R = ifelse(CHL > 16, 1, 0),
         AMP_R = ifelse(AMP > 8, 1, 0),
         GEN_R = ifelse(GEN > 2, 1, 0)) %>%
  select(contains("_R"), species) %>%
  split(., f = .$species) %>%
  lapply(., function(x) select(x, -species))

corr_matrices <- lapply(mic_per_species, function(x) create_corr_matrix(x)) %>%
  bind_rows(.id = "species")

cols <- c("#e7f0fa","#c9e2f6","#95cbee","#0099dc","#4ab04a",
          "#ffd73e","#eec73a","#e29421","#f05336","#ce472e")

ggplot(corr_matrices, aes(var1, var2, fill = value))+
  geom_tile(color = "white")+
  scale_fill_gradientn(colors=cols,
                       na.value="grey95",
                       limits=c(-1, 1),
                       guide=guide_colourbar(ticks=T,nbin=10,
                                             barheight=.5,label=T, 
                                             barwidth=10))+
  labs(fill = NULL)+
  theme_classic()+
  theme(axis.text = element_text(size = 7),
        axis.text.x = element_text(angle = 90, hjust = 1, vjust = 0.4),
        axis.title = element_blank(),
        legend.position="bottom",
        legend.justification="center",
        legend.direction="horizontal")+
  coord_fixed()+
  facet_wrap(~species)
```

## Minimum Spanning Tree of MLST profiles
Minimum spanning tree of sequence types, created in Phyloviz[3] with goeBURST distance. Light blue = Broiler, dark blue = Pig, light green = Red fox, and dark green = Wild bird. Distances between sequence types are denoted as numbers on the lines. 
![](../figures/mlst_tree_full.png)

## Heatmap plot
This plot represents the individual resistance mechanisms (x-axis) present in each isolate (y-axis), for each animal species (colors) and sequence type (dendrogram). Grey squares represent genes/mutations that ARIBA flagged as bad quality in their report. Dark color indicates presence, while the light color indicates absence. Part of the code below is taken from https://gist.github.com/scwatts/8c32d8f6dc44e352d89ae3de5f3a1125. The dendrogram was created by calculating the distance between each allele in the MLST scheme (Escherichia coli #1), and then use the UPGMA method to calculate the dendrogram.
```{r fig.height=30, fig.width=10, message=FALSE, warning=FALSE}
# Adjust MLST data frame
sequence_types_adjusted <- mlst_results %>%
  mutate_all(funs(gsub("\\*", "", .))) %>%
  select(-ST)

row.names(sequence_types_adjusted) <- sequence_types_adjusted$id

mlst <- sequence_types_adjusted[,-1]

# Calculate distance matrix from sequence typing alleles
d.mlst.distances <- matrix(0, ncol=nrow(mlst), nrow=nrow(mlst))
for (i in 1:(nrow(mlst)-1)) {
  for (j in (i+1):nrow(mlst)){
    d.mlst.distances[i,j] <- sum(mlst[i, ] != mlst[j, ])
    d.mlst.distances[j,i] <- sum(mlst[i, ] != mlst[j, ])
  }
}

# Create tree from distance matrix
tree <- upgma(d.mlst.distances)
id_label_tree <- tree 
id_label_tree$tip.label <- sequence_types_adjusted$id
tree$tip.label <- mlst_results$ST

# Create dummy plot of id_label_tree to extract correct order of ID's
p0 <- ggtree(id_label_tree)+
  geom_treescale()+
  geom_tiplab(align = TRUE, linesize = 0, hjust = -0.2, size = 1.5)+
  xlim(0,3.6)

# Get plot build data
p1b <- ggplot_build(p0)

# Save plot of tree
p1 <- ggtree(tree)+
  geom_treescale()+
  geom_tiplab(align = TRUE, linesize = 0, hjust = -0.2, size = 1.5)+
  xlim(0,3.6)

# Create heatmap data frame from gene data
megares_long <- megares_results %>%
  gather(gene, value, -c(id, type))

resfinder_long <- resfinder_results %>%
  gather(gene, value, -c(id, type))

heatmap_data <- megares_long %>%
  rbind(resfinder_long) %>%
  mutate(type2 = ifelse(type == "mut", 1, 2),
         value = factor(ifelse(is.na(value) == TRUE, 2, value))) %>%
  spread(id, value, fill = 0) %>%
  gather(key = "id", value = "result", -c(gene, type, type2)) %>%
  left_join(species_data, by = "id") %>%
  mutate(result = ifelse(result == 2, NA, result),
         fillcol = paste(species, result, sep = "_"),
         fillcol = ifelse(is.na(result) == TRUE, NA, fillcol))

# Arrange id-column after tip-labels in tree
heatmap_data <- heatmap_data %>%
  mutate(id = factor(id, levels = p1b$data[[5]] %>% arrange(y) %>% pull(label)))

# Color vector
palette <- c("Broiler_1" = "#a6cee3", "Broiler_0" = "#a6cee326",
             "Pig_1" = "#1f78b4", "Pig_0" = "#1f78b426",
             "Red fox_1" = "#b2df8a", "Red fox_0" = "#b2df8a26",
             "Wild bird_1" = "#33a02c", "Wild bird_0" = "#33a02c26")

# Plot the heatmap
p2 <- ggplot(heatmap_data, aes(fct_reorder(gene, type2),
             factor(id),
             fill = fillcol)) +
  geom_tile(color = "white")+
  geom_vline(xintercept = 13.5)+
  scale_fill_manual(values = palette,
                    na.value = "#6b6767")+
  theme_minimal()+
  labs(fill = NULL)+
  theme(axis.text.x = element_text(angle = 90,
                                   hjust = 1,
                                   vjust = 0.3,
                                   size = 7),
        axis.title = element_blank(),
        axis.text.y = element_blank(),
        panel.grid = element_blank(),
        legend.position = "right")+
  guides(fill = guide_legend(reverse = TRUE))+
  coord_fixed()

# Arrange both plots into one figure
grid.arrange(arrangeGrob(p1 + theme(plot.margin=margin(51,-2,135,0)),
                         heights = c(1.9,-0.097)),
             arrangeGrob(p2 + theme(axis.text.y=element_blank(),
                                    plot.margin = margin(0,0,0,-8)),
                         heights=c(0.98,-0.03)), 
             ncol=2)

```

# Statistics

## Significance test for the percent presence of acquired genes between animal species
Only significant values will be presented in the table below (p < 0.05).
```{r message=FALSE, warning=FALSE}
acquired_genes_list <- resfinder_results %>%
  left_join(isolate_data, by = "id") %>%
  gather(key, value, gene_names) %>%
  group_by(key, value, species) %>%
  count() %>%
  ungroup() %>%
  mutate(value = if_else(value == 1, "Present", "Absent")) %>%
  spread(value, n, fill = 0) %>%
  select(key, species, Present, Absent) %>%
  split(f = .$key)

p_values_acquired <- lapply(acquired_genes_list, function(x) do_chisq_test(x)) %>%
  bind_rows()

p_values_acquired
```

## Significance test for the percent presence of mutations in intrinsic genes between animal species
Only significant values will be presented in the table below (p < 0.05).
```{r message=FALSE, warning=FALSE}
mut_genes_list <- megares_results %>%
  left_join(isolate_data, by = "id") %>%
  gather(key, value, mut_names) %>%
  group_by(key, value, species) %>%
  count() %>%
  ungroup() %>%
  mutate(value = if_else(value == 1, "Present", "Absent")) %>%
  spread(value, n, fill = 0) %>%
  select(key, species, Present, Absent) %>%
  split(f = .$key)

p_values_mut <-  lapply(mut_genes_list, function(x) do_chisq_test(x)) %>%
  bind_rows()

p_values_mut
```

# References
[1] ARIBA: rapid antimicrobial resistance genotyping directly from sequencing reads Hunt M, Mather AE, Sánchez-Busó L, Page AJ, Parkhill J , Keane JA, Harris SR. Microbial Genomics 2017. doi:  https://doi.org/110.1099/mgen.0.000131

[2] ColorBrewer: An online tool for selecting color schemes for maps. The Cartographic Journal 40(1): 27-37

[3] PHYLOViZ: phylogenetic inference and data visualization for sequence based typing methods, Francisco A. P. et al.,  BMC Bioinformatics 2012, 13:87. doi: https://doi.org/10.1186/1471-2105-13-87